libs/tokens/src/hooks/tokens/useSearchToken.ts

```diff
@@ -0,0 +1,154 @@
+import { useMemo } from 'react'
+import { useAtomValue } from 'jotai'
+import { activeTokensAtom, inactiveTokensAtom } from '../../state/tokens/allTokensAtom'
+import { useDebounce } from '@cowprotocol/common-hooks'
+import { useWeb3React } from '@web3-react/core'
+import { isAddress } from '@cowprotocol/common-utils'
+import ms from 'ms.macro'
+import { getTokenSearchFilter } from '../../utils/getTokenSearchFilter'
+import useSWR from 'swr'
+import { searchTokensInApi } from '../../services/searchTokensInApi'
+import { TokenWithLogo } from '@cowprotocol/common-const'
+import { environmentAtom } from '../../state/environmentAtom'
+import { parseTokensFromApi } from '../../utils/parseTokensFromApi'
+import { fetchTokenFromBlockchain } from '../../utils/fetchTokenFromBlockchain'
+import { tokenWithLogoFromToken } from '../../utils/tokenWithLogoFromToken'
+
+const IN_LISTS_DEBOUNCE_TIME = ms`100ms`
+const IN_EXTERNALS_DEBOUNCE_TIME = ms`1s`
+
+export type TokenSearchResponse = {
+  isLoading: boolean
+  blockchainResult: TokenWithLogo[] | null
+  externalApiResult: TokenWithLogo[] | null
+  activeListsResult: TokenWithLogo[] | null
+  inactiveListsResult: TokenWithLogo[] | null
```

Comment:
do we need `null` state if we have empty list and `isLoading` state?

Reply:
Makes sense, thanks!
Refactored: https://github.com/cowprotocol/cowswap/pull/3201/commits/734c5f4db5445c5e118bf758f900a5d9828211b0